File: src/hackerrank/problem-solving/easy/missing-numbers/rb/missing_numbers_v1.rb
##
# Counts the frequencies in `xs`.
#
# @param xs [Array<Integer>
# @return [Hash]
#
def count_freqs(xs)
  freqs = {}

  xs.each do |x|
    freqs[x] = freqs[x] ? freqs[x] + 1 : 1
  end

  freqs
end

##
# Finds numbers that are present in `brr` but missing in `arr`.
#
# @param {number[]} arr
# @param {number[]} brr
# @returns {number[]} The array of the missing numbers (the
#   difference).
#
def missing_nums(arr, brr)
  freqs_arr = count_freqs(arr)
  freqs_brr = count_freqs(brr)
  missing = []

  freqs_brr.each_key do |key|
    missing << key if freqs_brr[key] != freqs_arr[key]
  end

  missing.sort
end

File: src/hackerrank/problem-solving/easy/missing-numbers/js/missing-numbers-v1.js
/**
 * Finds numbers that are present in `ys` but missing in `xs`.
 *
 * - T.C: O(n).
 * - S.C: O(n).
 *
 * @param {number[]} xs
 * @param {number[]} ys
 * @returns {number[]} The array of the missing numbers (the
 *   difference).
 */
function missingNums(arr, brr) {
  var freqXs = {},
      freqYs = {},
      i,
      n;

  for (i = 0; n = xs[i], i < xs.length; ++i)
    freqXs[n] = freqXs[n] + 1 || 1;

  for (i = 0; n = ys[i], i < ys.length; ++i)
    freqYs[n] = freqYs[n] + 1 || 1;

  return Object.keys(freqYs).reduce((missing, key) => {
    if (freqYs[key] === freqXs[key])
      return missing;

    missing.push(Number(key));

    return missing;
  }, []);
}

export { missingNums };
